wn/WDT

wakandanwilson opened the pull request:
## Change Description
Implemented hardware WDT to work with FreeRTOS.
## Why the change is needed and implementation details
Needed as a safety feature to ensure that all sensors are working, if not then we reset the teensy.
## Testing and validation
Tested by commenting out the Watchdog_Pet() line and added several print statements to track system flow. If Watchdog_Pet() is not called, the trigger statement will print to warn that system reset is happening and then once the timeout window closes, a statement will print and the system will reset automatically.

wakandanwilson committed:
updated vehicle-fw with WDT

billofright commented on fsae-vehicle-fw/src/main.cpp:
```diff
@@ -3,22 +3,132 @@
+  WDOG1_WSR = 0xAAAA;
+}
+
+bool checkSensors(){
```
I think checking each sensor is unnecessary. The analog signals will return a value regardless if there's a sensor fault because the pin is just reading the voltage. The IMU and GPS could time out but we can handle that in its own module.

billofright commented on fsae-vehicle-fw/src/main.cpp:
```diff
@@ -3,22 +3,132 @@
+
   while (true) {
     // Main loop
+    vTaskDelayUntil(&xLastWakeTime, xCheckInterval);
```
Can we use vTaskDelay instead of DelayUntil? I think its a bit more readable and straightforward.

billofright commented on fsae-vehicle-fw/src/main.cpp:
```diff
@@ -3,22 +3,132 @@
+  kickWatchdog();
+}
+
+void kickWatchdog(){
```
Can we call this Watchdog_Pet() LOL

billofright commented on fsae-vehicle-fw/src/main.cpp:
```diff
@@ -3,22 +3,132 @@
+void kickWatchdog();
+bool checkSensors();
+
+void initWatchdog(){
```
This looks great. Can we add all the watchdog functions into its own file? Probably make a watchdog.h and watchdog.cpp in the peripherals folder.

billofright reviewed (commented):
Really good start! A few structural things here and there but it's really good code. Try testing this on an actual teensy when you get the chance. Also, reminder to change the name of the PR to a relevant description!

billofright committed:
moved watchdog functions to new file and initd in main. needs to be tested on hardware

wakandanwilson committed:
moved watchdog functions into separate files. editted main.cpp to reflect these changes.

wakandanwilson committed:
separated watchdog functions into its own file. editted main.cpp to reflect these changes.

wakandanwilson committed:
fixed small error

wakandanwilson committed:
.pio\libdeps\teensy41\WDT_T4/Watchdog_t4.tpp:152:20: warning: there are no arguments to 'constrain' that depend on a template parameter, so a declaration of 'constrain' must be available [-fpermissive]
  152 |   config.trigger = constrain(config.trigger, 0.0f, 127.5); /* callback trigger before timeout */
      |                    ^~~~~~~~~
.pio\libdeps\teensy41\WDT_T4/Watchdog_t4.tpp:152:20: warning: there are no arguments to 'constrain' that depend on a template parameter, so a declaration of 'constrain' must be available [-fpermissive]
  152 |   config.trigger = constrain(config.trigger, 0.0f, 127.5); /* callback trigger before timeout */
      |                    ^~~~~~~~~# On branch wn/watchdog

wakandanwilson committed:
Merge branch 'main' of https://github.com/Anteater-Electric-Racing/embedded into wn/watchdog

wakandanwilson committed:
updates

wakandanwilson committed:
finished testing watchdog

billofright commented on fsae-vehicle-fw/src/embedded.code-workspace:
```diff
@@ -0,0 +1,16 @@
+{
```
what is this file

wakandanwilson replied:
im not sure why its there, it mightve been bc i saved it as a workspace. this file can be deleted.

billofright commented on fsae-vehicle-fw/src/main.cpp:
```diff
@@ -9,30 +9,34 @@
 
 void threadMain(void *pvParameters) {
     Serial.begin(9600);
+    vTaskDelay(pdMS_TO_TICKS(100));
```
each tick is by default 1ms so u dont have to do pdms_to_ticks

billofright commented on fsae-vehicle-fw/src/peripherals/watchdog.cpp:
```diff
@@ -0,0 +1,30 @@
+}
+
+void myCallBack(){
+  Serial.println("WARNING: WATCHDOG NEEDS TO BE PET.");
```
does it reboot the teensy?

wakandanwilson committed:
updated watchdog to have initialization checking and additional reset function for better readibility

ksthakkar commented:
2024-25 archive, closed
